ShareQueryCacheOverNodes option

SergeiPavlov opened the pull request:
Benchmarking:
```
|             Method |     Mean | Error |      Min |      Max |  Gen 0 | Allocated |
|------------------- |---------:|------:|---------:|---------:|-------:|----------:|
|   SharedQueryCache | 208.8 us |    NA | 208.8 us | 208.8 us |      - |     27 KB |
| SeparateQueryCache | 825.9 us |    NA | 825.9 us | 825.9 us | 1.0000 |    256 KB |
```
10 Nodes
QueryCacheSize: 30
25 various queries

SergeiPavlov committed:
ShareQueryCacheOverNodes option

botinko commented on Orm/Xtensive.Orm/Core/ParameterContext.cs:
```diff
@@ -47,4 +54,17 @@ public ParameterContext(ParameterContext outerContext = null)
     }
   }
+
+  public class TypeIdParameterContext : ParameterContext
```
sealed?

SergeiPavlov replied:
done

botinko commented on Orm/Xtensive.Orm/Orm/Internals/CompiledQueryRunner.cs:
```diff
@@ -212,13 +212,13 @@ private ParameterContext CreateParameterContext(ParameterizedQuery query)
     }
 
-    public CompiledQueryRunner(QueryEndpoint endpoint, object queryKey, object queryTarget, ParameterContext outerContext = null)
+    public CompiledQueryRunner(QueryEndpoint endpoint, object queryKey, object queryTarget, ParameterContext outerContext = null, bool sharedQueryCache = false)
```
default value for sharedQueryCache make it easy to forget to pass proper sharedQueryCache value from session. 
Can we treat sharedQueryCache  as endpoint.Provider.Session.Domain.Configuration.ShareQueryCacheOverNodes ? It will avoid errors and will make code cleaner.

SergeiPavlov replied:
done

botinko commented on Orm/Xtensive.Orm/Sql/Compiler/SqlCompilationResult.cs:
```diff
@@ -17,6 +17,8 @@ public sealed class SqlCompilationResult
     private readonly IDictionary<object, string> parameterNames;
     private volatile int lastResultLength;
 
+    public Dictionary<object, string> PlaceholderValues;
```
Look like SqlCompilationResult is an object, which was created fairly often. We can safe a bit resources by making PlaceholderValues a simple `List<Pair<object, strinig>>`, as it will be few `PlaceholderValues` actually.

botinko replied:
looks like should be private

botinko commented on Orm/Xtensive.Orm/Sql/Compiler/SqlCompilationResult.cs:
```diff
@@ -17,6 +17,8 @@ public sealed class SqlCompilationResult
     private readonly IDictionary<object, string> parameterNames;
     private volatile int lastResultLength;
 
+    public Dictionary<object, string> PlaceholderValues;
```
I can't find any usages of that field.

botinko commented on Orm/Xtensive.Orm/Sql/Compiler/Internals/PostCompiler.cs:
```diff
@@ -21,9 +21,14 @@ internal class PostCompiler : NodeVisitor
     private string[] currentCycleItem;
 
-    public static string Process(Node root, SqlPostCompilerConfiguration configuration, int estimatedResultLength)
+    public static string Process(Node root, SqlPostCompilerConfiguration configuration, int estimatedResultLength, Dictionary<object, string> placeholders)
```
looks like placeholders is always null.

botinko commented on Orm/Xtensive.Orm/Sql/Compiler/SqlCompilationResult.cs:
```diff
@@ -17,6 +17,8 @@ public sealed class SqlCompilationResult
     private readonly IDictionary<object, string> parameterNames;
     private volatile int lastResultLength;
 
+    public Dictionary<object, string> PlaceholderValues;
```
I'd make it private and initialize from ctor.

SergeiPavlov replied:
It is optional property. In most cases is `null`

SergeiPavlov committed:
minor changes

SergeiPavlov committed:
SqlCompilationResult.placeholderValues private property

AlexUstinov commented on Orm/Xtensive.Orm.SqlServer/Sql.Drivers.SqlServer/v09/Translator.cs:
```diff
@@ -261,8 +261,9 @@ private static SqlJoinHint TryFindJoinHint(SqlCompilerContext context, SqlJoinEx
       for (int i = 0, count = context.GetTraversalPath().Length; i < count; i++) {
-        if (context.GetTraversalPath()[i] is SqlQueryStatement)
-          statement = context.GetTraversalPath()[i] as SqlQueryStatement;
+        if (context.GetTraversalPath()[i] is SqlQueryStatement sqlQueryStatement) {
```
It also makes sense to go farther and have a traversal path stored into a variable before running the `for` loop. e.g.
```csharp
  var traversalPath = context.GetTraversalPath();
  for (int i = 0, count = traversalPath.Length; i < count; i++) {
    if (traversalPath[i] is SqlQueryStatement sqlQueryStatement) {
     // ...
    }
  }
    
```

SergeiPavlov replied:
done

AlexUstinov commented on Orm/Xtensive.Orm.SqlServer/Sql.Drivers.SqlServer/v09/Translator.cs:
```diff
@@ -261,8 +261,9 @@ private static SqlJoinHint TryFindJoinHint(SqlCompilerContext context, SqlJoinEx
-        if (context.GetTraversalPath()[i] is SqlQueryStatement)
-          statement = context.GetTraversalPath()[i] as SqlQueryStatement;
+        if (context.GetTraversalPath()[i] is SqlQueryStatement sqlQueryStatement) {
+          statement = sqlQueryStatement;
```
Should we go the `traversalPath` in the opposite direction and break after we find the first `SqlQueryStatement` occurrence?

SergeiPavlov replied:
done

AlexUstinov commented on Orm/Xtensive.Orm/Orm/Building/Builders/ProviderDescriptor.cs:
```diff
@@ -75,14 +75,17 @@ private static Assembly LoadExtensionAssembly(string extensionAssemblyName)
           Strings.ExAssemblyVersionMismatchMainAssemblyXYExtensionsAssemblyAB,
           mainAssemblyRef.Name, mainAssemblyVersion,
           extensionAssemblyName, extensionAssemblyVersion));
+      */
```
Is there an intention to return this file to its initial state?

SergeiPavlov replied:
Yes, I missed this temporary local change

AlexUstinov commented on Orm/Xtensive.Orm/Core/ParameterContext.cs:
```diff
@@ -47,4 +54,17 @@ public ParameterContext(ParameterContext outerContext = null)
     }
   }
+
+  public sealed class TypeIdParameterContext : ParameterContext
```
Can we instead of inheritance have the `TypeIdRegistry` field in the `ParameterContext` class? The default value would be `null` and we could return a result of the parent's context `GetTypeId` method call in that case.

SergeiPavlov replied:
done, thanks

AlexUstinov commented on Orm/Xtensive.Orm/Orm/Providers/SqlCompiler.Index.cs:
```diff
@@ -5,18 +5,54 @@
+    protected readonly struct QueryAndBindings
+    {
+      public SqlSelect Query { get; }
+      public List<QueryParameterBinding> Bindings { get; }
```
I believe the `List<>` type is not a proper choice in this case. I'd prefer to see `IReadOnlyCollection<>` here.

SergeiPavlov replied:
done

AlexUstinov commented on Orm/Xtensive.Orm/Orm/Providers/SqlCompiler.Index.cs:
```diff
@@ -31,11 +67,11 @@ protected override SqlProvider VisitContainsTable(ContainsTableProvider provider
-      SqlSelect query = BuildProviderQuery(index);
-      return CreateProvider(query, provider);
+      var queryAndBindings = BuildProviderQuery(index);
+      return CreateProvider(queryAndBindings.Query, queryAndBindings.Bindings.ToArray(), provider);
```
It is better to avoid the `ToArray` call here.

SergeiPavlov replied:
done

AlexUstinov commented on Orm/Xtensive.Orm/Sql/Compiler/Internals/PostCompiler.cs:
```diff
@@ -21,9 +21,14 @@ internal class PostCompiler : NodeVisitor
     private string[] currentCycleItem;
 
-    public static string Process(Node root, SqlPostCompilerConfiguration configuration, int estimatedResultLength)
+    public static string Process(Node root, SqlPostCompilerConfiguration configuration, int estimatedResultLength, Dictionary<object, string> placeholders)
```
Can we change the parameter type to the `IReadOnlyDictionary<,>`?

SergeiPavlov replied:
done

AlexUstinov commented on Orm/Xtensive.Orm/Sql/Compiler/SqlCompilationResult.cs:
```diff
@@ -17,6 +17,8 @@ public sealed class SqlCompilationResult
     private readonly IDictionary<object, string> parameterNames;
     private volatile int lastResultLength;
 
+    private readonly Dictionary<object, string> placeholderValues;
```
This field may be of the `IReadOnlyDictionary<,>` type.

SergeiPavlov replied:
done

AlexUstinov commented on Orm/Xtensive.Orm/Sql/Compiler/SqlCompilationResult.cs:
```diff
@@ -60,15 +62,15 @@ public string GetCommandText(SqlPostCompilerConfiguration configuration)
     // Constructors
 
-    internal SqlCompilationResult(Node result, IDictionary<object, string> parameterNames)
+    internal SqlCompilationResult(Node result, IDictionary<object, string> parameterNames, Dictionary<object, string> placeholderValues)
```
`IReadOnlyDictionary<,>`

SergeiPavlov replied:
done

AlexUstinov commented on Orm/Xtensive.Orm/Sql/Compiler/SqlCompiler.cs:
```diff
@@ -191,17 +195,27 @@ public virtual void Visit(SqlAlterSequence node)
-        context.Output.AppendText(translator.Translate(context, items[i]));
-        context.Output.AppendDelimiter(translator.RowItemDelimiter);
+      if (node.ItemType == SqlPlaceholderType) {
+        for (var i = 0; i < items.Length - 1; i++) {
```
We could store `items.Length` to a variable and prevent repeated property access during iteration.

SergeiPavlov replied:
done

AlexUstinov commented on Orm/Xtensive.Orm/Sql/Compiler/SqlCompiler.cs:
```diff
@@ -191,17 +195,27 @@ public virtual void Visit(SqlAlterSequence node)
+        context.Output.AppendText(translator.Translate(context, node, ArraySection.Exit));
+      }
+      else {
+        for (var i = 0; i < items.Length - 1; i++) {
```
The same is here.

SergeiPavlov replied:
done

AlexUstinov commented on Orm/Xtensive.Orm/Sql/Compiler/SqlCompiler.cs:
```diff
@@ -1353,9 +1367,17 @@ public virtual void Visit(SqlTableColumn node)
+      var text = translator.Translate(context, node, TableSection.Entry) +
+        translator.Translate(context, node, TableSection.AliasDeclaration);
+      const string placeholder = "[node_placeholder]";
+      if (text.IndexOf(placeholder) is var idx && idx < 0) {
```
It might be a good idea to use the `StringComparison.Ordinal` here.

SergeiPavlov replied:
done

AlexUstinov commented on Orm/Xtensive.Orm/Sql/Compiler/SqlTranslator.cs:
```diff
@@ -1552,6 +1552,12 @@ public virtual string Translate(SqlCompilerContext context, SchemaNode node)
       var actualizer = context.SqlNodeActualizer;
 
+      if (!dbQualified && context.ShareQueryCacheOverNodes) {
+        var id = "node_placeholder";
```
`const string` ?

SergeiPavlov replied:
done

SergeiPavlov committed:
Minor improvements

SergeiPavlov commented on Orm/Xtensive.Orm.SqlServer/Sql.Drivers.SqlServer/v09/Translator.cs:
```diff
@@ -548,7 +549,7 @@ public override string Translate(SqlCompilerContext context, object literalValue
     public override string Translate(SqlLockType lockType)
     {
-      var items = new List<string>();
+      var items = new List<string>(5);
```
changed to `3`

AlexUstinov commented on Orm/Xtensive.Orm/Sql/Compiler/Internals/PostCompiler.cs:
```diff
@@ -21,12 +21,14 @@ internal class PostCompiler : NodeVisitor
       if (placeholders != null) {
         foreach (var kv in placeholders) {
+#pragma warning disable IDE0058 // Expression value is never used
           compiler.configuration.PlaceholderValues.TryAdd(kv.Key, kv.Value);  // Don't override values added in CreateQueryPart()
```
Can we use 
```csharp
  _ = compiler.configuration.PlaceholderValues.TryAdd(kv.Key, kv.Value);
```
instead of pragma?

SergeiPavlov replied:
done

AlexUstinov commented on Orm/Xtensive.Orm/Sql/Compiler/SqlTranslator.cs:
```diff
@@ -1553,8 +1553,10 @@ public virtual string Translate(SqlCompilerContext context, SchemaNode node)
-        var id = "node_placeholder";
+        const string id = "node_placeholder";
+#pragma warning disable IDE0058 // Expression value is never used
         context.PlaceholderValues.TryAdd($"[{id}]", QuoteIdentifier(actualizer.Actualize(node.Schema)));
```
The same is here, we would benefit of `_ = ...` syntax.

botinko replied:
inline [node_placeholder]

SergeiPavlov replied:
> inline [node_placeholder]

The nex line uses `id` too:
`return QuoteIdentifier(id`

SergeiPavlov replied:
Refactored to static:
```
    private const string NodePlaceholder = "node_placeholder";
    public const string BracketedNodePlaceholder = "[node_placeholder]";
```

botinko commented on Orm/Xtensive.Orm/Core/ParameterContext.cs:
```diff
@@ -7,22 +7,28 @@
   /// Provides storing context-specific <see cref="Parameter{TValue}"/>'s values.
   /// </summary>
-  public sealed class ParameterContext
+  public class ParameterContext
```
sealed

SergeiPavlov replied:
done

botinko commented:
` ShareQueryCacheOverNodes`  doesn't reflect right meaning. QueryCache is already shared between nodes. Maybe something like `ShareCompiledQueriesOverDomain` fits better? @AlexUstinov what do you think?

botinko commented on Orm/Xtensive.Orm/Sql/Compiler/SqlTranslator.cs:
```diff
@@ -1552,6 +1552,14 @@ public virtual string Translate(SqlCompilerContext context, SchemaNode node)
       var actualizer = context.SqlNodeActualizer;
 
+      if (!dbQualified && context.ShareQueryCacheOverNodes) {
+        const string id = "node_placeholder";
```
I'd move `[node_placeholder`]` constant to global static, as it used in 3 places independently now.

SergeiPavlov replied:
done

botinko commented on Orm/Xtensive.Orm/Sql/Compiler/SqlCompiler.cs:
```diff
@@ -1353,9 +1368,17 @@ public virtual void Visit(SqlTableColumn node)
-      context.Output.AppendText(
-        translator.Translate(context, node, TableSection.Entry)+
-        translator.Translate(context, node, TableSection.AliasDeclaration));
+      var text = translator.Translate(context, node, TableSection.Entry) +
```
Isn't it a good idea to scan for placeholder only `translator.Translate(context, node, TableSection.Entry)`?

SergeiPavlov replied:
Optimized

botinko commented on Orm/Xtensive.Orm.Tests/Storage/CommandProcessing/BatchingCommandProcessorParametersManagement.cs:
```diff
@@ -45,6 +44,11 @@ protected override void PopulateData()
+
+      maxQueryParameterCount = StorageProviderInfo.Instance.Info.MaxQueryParameterCount;
+
+      parametersFitInSingleCommand = Enumerable.Range(1, maxQueryParameterCount - 1).ToArray(maxQueryParameterCount - 1);
```
`maxQueryParameterCount - 1` parameter is not actually needed, as ToArray takes into account Enumerator actual type to recognize resulted array size. It's from dotnet 3

SergeiPavlov replied:
fixed

botinko commented on Orm/Xtensive.Orm/Orm/Providers/SqlCompiler.Index.cs:
```diff
@@ -82,40 +118,67 @@ private SqlSelect BuildTableQuery(IndexInfo index)
-          var tableRef = (SqlTableRef) q.From;
-          return (SqlTable) SqlDml.TableRef(tableRef.DataTable, tableRef.Name, q.Columns.Select(c => c.Name));
-        }).ToList();
+      var sourceTables = new List<SqlTable>();
```
Please, see my patch for this method (in DM)

SergeiPavlov replied:
done

SergeiPavlov committed:
Improve LINQ and allocations

SergeiPavlov committed:
Minor optimizations

AlexUstinov commented on Orm/Xtensive.Orm/Orm/Providers/SqlCompiler.Index.cs:
```diff
@@ -152,31 +152,27 @@ private QueryAndBindings BuildJoinQuery(IndexInfo index)
-      var underlyingQueries = index.UnderlyingIndexes.Select(BuildProviderQuery);
+      var underlyingQueries = new QueryAndBindings[index.UnderlyingIndexes.Count];
+      var haveVirtualUnderlyingIndexes = false;
+      for (var i = 0; i < index.UnderlyingIndexes.Count; i++) {
```
We have repeated access in the loop to the `UnderlyingIndexes` and the `Count` properties.

SergeiPavlov replied:
fixed

AlexUstinov commented on Orm/Xtensive.Orm/Orm/Providers/SqlCompiler.Index.cs:
```diff
@@ -152,31 +152,27 @@ private QueryAndBindings BuildJoinQuery(IndexInfo index)
-            resultBindings.AddRange(item.Bindings);
-          }
+
+      for (var i = 0; i < underlyingQueries.Length; i++) {
```
Here we have repeated access to the `Length` property.

SergeiPavlov replied:
fixed

AlexUstinov commented on Orm/Xtensive.Orm/Orm/Providers/SqlCompiler.Index.cs:
```diff
@@ -218,6 +214,18 @@ private QueryAndBindings BuildJoinQuery(IndexInfo index)
+    {
+      var columns = item.Query.Columns;
+      var columnNames = new string[columns.Count];
+      for (var i = 0; i < columns.Count; i++) {
```
Repeated access to the `Count` property.

SergeiPavlov replied:
fixed

SergeiPavlov committed:
Some refactoring

SergeiPavlov commented:
> ` ShareQueryCacheOverNodes` doesn't reflect right meaning. QueryCache is already shared between nodes. Maybe something like `ShareCompiledQueriesOverDomain` fits better? @AlexUstinov what do you think?

I'm ready change the name. Lets make consensus about it

SergeiPavlov committed:
small fix

SergeiPavlov committed:
Refactor [node_placeholder] string

SergeiPavlov committed:
small optimizations

SergeiPavlov committed:
small optimizations

AlexUstinov commented on Orm/Xtensive.Orm.Tests/Storage/CommandProcessing/BatchingCommandProcessorParametersManagement.cs:
```diff
@@ -47,8 +47,8 @@ protected override void PopulateData()
 
-      parametersFitInSingleCommand = Enumerable.Range(1, maxQueryParameterCount - 1).ToArray(maxQueryParameterCount - 1);
-      parametersDontFitInSingleCommand = Enumerable.Range(1, maxQueryParameterCount + 1).ToArray(maxQueryParameterCount + 1);
+      parametersFitInSingleCommand = Enumerable.Range(1, maxQueryParameterCount - 1).ToArray();
```
Why? Don't we benefit from pre-allocating array of the needed length?

botinko replied:
from net core 3.1 ToArray checked actual type of IEnumerable, and if it is if it's IIListProvider<T> use it for array allocation. 
it works even in this scenario: `someEnumerable.OrderBy(i => i.SomeField).Take(1)` will not sort all items, but will take Min (this is from .net 6 afaik).

AlexUstinov replied:
It makes sense, then.

AlexUstinov commented on Orm/Xtensive.Orm/Orm/Providers/SqlCompiler.Index.cs:
```diff
@@ -231,7 +232,7 @@ private QueryAndBindings BuildFilteredQuery(IndexInfo index)
       var baseQueryAndBindings = BuildProviderQuery(underlyingIndex);
       var baseQuery = baseQueryAndBindings.Query;
-      var bindings = baseQueryAndBindings.Bindings;
+      var bindings = baseQueryAndBindings.Bindings.ToList();
```
This must be considered as a place for future optimizations. In many cases this bindings list stays unchanged but we always copying it.

SergeiPavlov replied:
optimized

SergeiPavlov committed:
Refactor to use SqlCompiler.AppendTranslated()

AlexUstinov commented on Orm/Xtensive.Orm/Sql/Compiler/SqlCompiler.cs:
```diff
@@ -211,32 +216,32 @@ public virtual void Visit(SqlArray node)
         for (int i = 0; i < lengthMinusOne; i++) {
-          output.AppendText(translator.Translate(context, items[i]));
+          AppendTranslatedLiteral(items[i]);
           output.AppendDelimiter(translator.RowItemDelimiter);
```
Shouldn't we have a wrapper method for the `AppendDelimiter` as well? Otherwise, it doesn't look uniform.

SergeiPavlov replied:
Yes, but I'm going to include  `AppendDelimiter` together with this last commint 
into other independed PR.
Bacause such an optimization would be useful even without SharedQueryCache

SergeiPavlov replied:
refactored it

commit-n-run commented on Orm/Xtensive.Orm/Sql/Compiler/SqlTranslator.cs:
```diff
@@ -122,12 +125,12 @@ public virtual string Translate(SqlCompilerContext context, SqlAlterDomain node,
     {
       var builder = new StringBuilder();
-      builder.Append("ALTER PARTITION FUNCTION " + QuoteIdentifier(node.PartitionFunction.DbName) + "()");
+      _ = builder.Append("ALTER PARTITION FUNCTION " + QuoteIdentifier(node.PartitionFunction.DbName) + "()");
```
Why do we need `_ =` here?
It's typical to ignore `StringBuilder.Append()` result

SergeiPavlov replied:
> Why do we need `_ =` here?
> It's typical to ignore `StringBuilder.Append()` result

 warning `IDE0058: computed value is never used`.  The ISE underlines such lines
Alexey Kulakov also asks me to fix such cases in modified code.

SergeiPavlov replied:
removed `_`

commit-n-run commented on Orm/Xtensive.Orm/Sql/Compiler/SqlCompilerConfiguration.cs:
```diff
@@ -29,6 +29,8 @@ public sealed class SqlCompilerConfiguration
     /// </summary>
     public bool DatabaseQualifiedObjects { get; set; }
 
+    public bool ShareQueryCacheOverNodes { get; set; }
```
It's better to name this option something like `ParametrizeSchemaNames`.
`SqlCompiler` does not know anything about query cache or nodes.

SergeiPavlov replied:
@botinko  suggested other name like `ShareQueriesOverDomain`
Lets make consensus

commit-n-run replied:
I'm not talking about `DomainConfiguration.ShareQueryCacheOverNodes`
I'm talking about `SqlCompilerConfiguration.ShareQueryCacheOverNodes`
`SqlCompiler` does not have concept of _nodes_ or _query cache_.
So at `SqlCompiler` level `ShareQueryCacheOverNodes` makes no sense

SergeiPavlov replied:
renamed to `ParametrizeSchemaNames`

commit-n-run commented on Orm/Xtensive.Orm/Sql/Compiler/SqlCompilerContext.cs:
```diff
@@ -16,12 +16,16 @@ public class SqlCompilerContext
     private readonly Stack<SqlNode> traversalStack;
     private readonly HashSet<SqlNode> traversalTable;
 
+    public bool ShareQueryCacheOverNodes { get; set; }
```
Same here

SergeiPavlov replied:
Renamed to  `ParametrizeSchemaNames`

commit-n-run commented on Orm/Xtensive.Orm/Orm/Providers/SqlCompiler.Index.cs:
```diff
@@ -5,18 +5,54 @@
+
+    private static readonly Func<TypeInfo, Func<ParameterContext, object>> ContextBasedAccessorFactory = type => ctx => ctx.GetTypeId(type);
+
+    private static readonly ConcurrentDictionary<int, Func<ParameterContext, object>> typeIdParameterAccessorCache2 =
```
We can have multiple `Domain` instances with different configuration.
Those might have overlapping type id values.
Caching statically will break this.
You can create separate type for these caches and create once instance of it per domain

SergeiPavlov replied:
removed this code

commit-n-run commented on Orm/Xtensive.Orm/Sql/Compiler/Internals/PostCompiler.cs:
```diff
@@ -21,9 +21,14 @@ internal class PostCompiler : NodeVisitor
     private string[] currentCycleItem;
 
-    public static string Process(Node root, SqlPostCompilerConfiguration configuration, int estimatedResultLength)
+    public static string Process(Node root, SqlPostCompilerConfiguration configuration, int estimatedResultLength, IReadOnlyDictionary<object, string> placeholders)
```
It looks odd that we have additional `placeholders` parameter even we already have `PlaceholderValues` in `SqlPostCompilerConfiguration`

SergeiPavlov replied:
fixed it

commit-n-run commented on Orm/Xtensive.Orm/Sql/Compiler/Internals/PostCompiler.cs:
```diff
@@ -21,9 +21,14 @@ internal class PostCompiler : NodeVisitor
       var compiler = new PostCompiler(configuration, estimatedResultLength);
+      if (placeholders != null) {
+        foreach (var kv in placeholders) {
+          _ = compiler.configuration.PlaceholderValues.TryAdd(kv.Key, kv.Value);  // Don't override values added in CreateQueryPart()
```
Generally it's assumed that we don't mutate user configuration objects

SergeiPavlov replied:
I could not find better place to inject this placeholder mapping

SergeiPavlov replied:
removed this code

commit-n-run commented on Orm/Xtensive.Orm/Sql/Compiler/SqlTranslator.cs:
```diff
@@ -1552,6 +1555,11 @@ public virtual string Translate(SqlCompilerContext context, SchemaNode node)
       var actualizer = context.SqlNodeActualizer;
 
+      if (!dbQualified && context.ShareQueryCacheOverNodes) {
+        _ = context.PlaceholderValues.TryAdd(BracketedNodePlaceholder, QuoteIdentifier(actualizer.Actualize(node.Schema)));
```
We are using the key key for all schemes, generally DO support multiple schemes in single query and this will break.

SergeiPavlov replied:
Placeholder key changed to `Schema` object.

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Merge remote-tracking branch 'origin/master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
rename SqlCompilerConfiguration.ShareQueryCacheOverNodes -> .ParametrizeSchemaNames

SergeiPavlov committed:
Substitute TypeId by PlaceholderNode

SergeiPavlov committed:
Remove doubling PlaceholderValues arg

SergeiPavlov committed:
Optimize bindings.ToList()

SergeiPavlov committed:
Remove TypeIdRegistry from SqlCompiler

SergeiPavlov committed:
Enclose schame into Opener/Closer chars

SergeiPavlov committed:
Get rid of unused TypeIdRegistry fields

SergeiPavlov committed:
Optimize serializing TypeId

SergeiPavlov committed:
Add comment into AppendPlaceholderValue()

SergeiPavlov committed:
Fix TypeId substitution for (x is Type) expressions

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Avoid strong version check during debugging

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Optimize GetMemberType() extension; TypeHelper

SergeiPavlov committed:
Merge remote-tracking branch 'origin/master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Fix NRE in SqlCompiler when ShareStorageSchemaOverNodes==false

SergeiPavlov committed:
Fix NRE in SqlCompiler when ShareStorageSchemaOverNodes==false

SergeiPavlov committed:
Check ShareQueryCacheOverNodes / ShareStorageSchemaOverNodes compatibility in Domain.Build()

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'master-servicetitan' into Optimize_GetMemberType

SergeiPavlov committed:
Rollback  GetGenericInterface() optimization as insignificant

SergeiPavlov committed:
Change GetMemberType() to use MetadataToken

SergeiPavlov committed:
Revert "Change GetMemberType() to use MetadataToken"

This reverts commit c55b013541a5b6106ab6a5a5c5ee20b2892fac6f.

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Memoize MakeGenericMethod()

SergeiPavlov committed:
Memoize MakeGenericType()

SergeiPavlov committed:
Improve readability

SergeiPavlov committed:
Get rid of TypeIdRegistry from SqlCompiler

SergeiPavlov committed:
Get rid of TypeIdRegistry from SqlCompiler

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Revert "Get rid of TypeIdRegistry from SqlCompiler"

This reverts commit a1922c34ba9248187efe36a3349892056a49448e.

SergeiPavlov committed:
Merge branch 'Optimize_GetMemberType' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'MakeGenericMethod' into SharedQueryCache

AlexUstinov commented on Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs:
```diff
@@ -58,20 +58,24 @@ protected override Expression VisitTypeIs(TypeBinaryExpression tb)
+
+        // Cast to TypeInfo to be compatible with TypeInfo literals
+        // The literals are replaced to placeholders in SqlTranslator
+        memberExpression = Expression.TypeAs(memberExpression, typeof(TypeInfo));
```
What is TypeInfo literal?

SergeiPavlov replied:
TypeInfo liters were converted to Placeholders(TypeInfo) to avoid numeric type identifiers.
Unfortunately thi sapproach does not work for materializer

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Check context.Translator.state.BuildingProjection

SergeiPavlov committed:
Check itemMaterializationContextParameter

SergeiPavlov committed:
Check for typeId == 0

SergeiPavlov committed:
Check for typeId == TypeInfo.NoTypeId

SergeiPavlov committed:
use nameof() for method names

SergeiPavlov committed:
use nameof() for method names

SergeiPavlov committed:
use nameof() for method names

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Optimize size of PackedFieldDescriptor struct from 16 bytes to 8

Increment copyright year

Avoid recalculating .ValueIndex/.ValueBitOffset

avoid double array indexing by using `ref var`

avoid double array indexing by using `ref var`

Calc ValueFieldAccessor.Rank() using BitOperations.Log2()

avoid double array indexing by using `ref var`

avoid double array indexing by using `ref var`

Optimize PackedTuple.GetHashCode()

Use 'in' instead of 'ref' where possible

Use 'in' instead of 'ref' where possible

avoid getting .Length/.Count property on each iteration

avoid getting .Length/.Count property on each iteration

Refactor TupleExtensions.CopyTo()

Renumber PackedFieldAccessor indexes to avoid gaps

SergeiPavlov committed:
Make .ForEach() extension [Obsolete]; refactor to foreach

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Bump version

SergeiPavlov committed:
DO_TargetFrameworks env var to multitargeting build

SergeiPavlov committed:
Fix ambiguity in initializin WellKnownMembers.Queryable members because of new IQueryable extension overloads in .NET6

SergeiPavlov committed:
Fix ambiguity in initializin WellKnownMembers.Queryable members because of new IQueryable extension overloads in .NET6

SergeiPavlov committed:
Bump version

SergeiPavlov committed:
Fix MaxWithSelector, MinWithSelector initializing

SergeiPavlov committed:
Fix MaxWithSelector, MinWithSelector initializing

SergeiPavlov committed:
Optimization: avoid Lambda JITing for typical case of invoking Bulk .Set() with value

SergeiPavlov committed:
Clear temporary tables by TRUNCATE TABLE statement for efficiency

SergeiPavlov committed:
Merge branch 'Remove_ForEach' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'upstream/optimize_bulk' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'upstream/fix_WellKnownMembers.Queryable' into SharedQueryCache

SergeiPavlov committed:
Bump version

SergeiPavlov committed:
Merge branch 'upstream/PackedDescriptor_ref_readonly' into SharedQueryCache

SergeiPavlov committed:
Fix Bulk set operation for entities

SergeiPavlov committed:
Bump version

SergeiPavlov committed:
Upgrade BitFaster.Caching

SergeiPavlov committed:
Update WellKnown.cs

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'upstream/multitarget_build' into SharedQueryCache

SergeiPavlov committed:
After-merge fix

SergeiPavlov committed:
Don't use $TargetFramework in DataObjects.Net.targets

SergeiPavlov committed:
Optimization: make event arguments readonly structs

SergeiPavlov committed:
Refactor event invocations

SergeiPavlov committed:
Rollback QueryEventArgs to class

SergeiPavlov committed:
Make MaxNumberOfConditions configurable

SergeiPavlov committed:
Merge branch 'upstream/SqlTruncateTable' into SharedQueryCache

SergeiPavlov committed:
XtensiveWeaverFramework property

SergeiPavlov committed:
Bump version

SergeiPavlov committed:
Make .Chunk() extension internal

SergeiPavlov committed:
Make .Chunk() extension internal

SergeiPavlov committed:
Make ArgumentValidator.EnsureArgumentNotNull() obsolete for .NET6

SergeiPavlov committed:
Optimize SqlFunctionCall/SqlCustomFunctionCall

SergeiPavlov committed:
Optimize logging

SergeiPavlov committed:
Merge branch 'upstream/optimize_Logging' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'upstream/struct_EventArgs' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
bump version

SergeiPavlov committed:
Optimizing refactoring: convert EnsureNotLocked() from extension to Method

SergeiPavlov committed:
Merge branch 'master-servicetitan' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'upstream/EnsureNotLocked2' into SharedQueryCache

SergeiPavlov committed:
Merge branch 'upstream/Optimize_FunctionCall' into SharedQueryCache

SergeiPavlov committed:
Update version

SergeiPavlov committed:
Implement Lazy 2-level multi-row INSERT batches: sizes 16, 256

SergeiPavlov committed:
Make 1-row INSERT Lazy

SergeiPavlov committed:
Make ClearRequest lazy

SergeiPavlov committed:
Merge branch 'upstream/SqlTruncateTable' into SharedQueryCache

SergeiPavlov committed:
Bump version
